tron/contracts/factory/Factory.sol

```diff
@@ -0,0 +1,394 @@
+pragma solidity ^0.4.24;
+
+import "../utils/OwnableContract.sol";
+import "../controller/ControllerInterface.sol";
+
+
+contract Factory is OwnableContract {
+
+    enum RequestStatus {PENDING, CANCELED, APPROVED, REJECTED}
+
+    struct Request {
+        address requester; // sender of the request.
+        uint amount; // amount of token to mint/burn.
+        string depositAddress; // custodian's asset address in mint, merchant's asset address in burn.
+        string txid; // asset txid for sending/redeeming asset in the mint/burn process.
+        uint nonce; // serial number allocated for each request.
+        uint timestamp; // time of the request creation.
+        RequestStatus status; // status of the request.
+    }
+
+    ControllerInterface public controller;
+
+    // mapping between merchant to the corresponding custodian deposit address, used in the minting process.
+    // by using a different deposit address per merchant the custodian can identify which merchant deposited.
+    mapping(address=>string) public custodianDepositAddress;
+
+    // mapping between merchant to the its deposit address where the asset should be moved to, used in the burning process.
+    mapping(address=>string) public merchantDepositAddress;
+
+    // mapping between a mint request hash and the corresponding request nonce.
+    mapping(bytes32=>uint) public mintRequestNonce;
+
+    // mapping between a burn request hash and the corresponding request nonce.
+    mapping(bytes32=>uint) public burnRequestNonce;
+
+    Request[] public mintRequests;
+    Request[] public burnRequests;
+
+    constructor(ControllerInterface _controller) public {
+        require(_controller != address(0), "invalid _controller address");
+        controller = _controller;
+        owner = _controller;
+    }
+
+    modifier onlyMerchant() {
+        require(controller.isMerchant(msg.sender), "sender not a merchant.");
+        _;
+    }
+
+    modifier onlyCustodian() {
+        require(controller.isCustodian(msg.sender), "sender not a custodian.");
+        _;
+    }
+
+    event CustodianDepositAddressSet(address indexed merchant, address indexed sender, string depositAddress);
+
+    function setCustodianDepositAddress(
+        address merchant,
+        string depositAddress
+    )
+        external
+        onlyCustodian
+        returns (bool)
+    {
+        require(merchant != 0, "invalid merchant address");
+        require(controller.isMerchant(merchant), "merchant address is not a real merchant.");
+        require(!isEmptyString(depositAddress), "invalid asset deposit address");
```

Comment:
We seem to be validating addresses quite a bit in this contract.  Gas permitting we may want to drop this into its own internal function:

`validateAddress(depositAddress)`

Reply:
yeah this is not a bad idea. I don't want to mutate the underlying thing too much because of the existing audit but this is not bad feedback at all